libs/sdk-ui-ext/src/internal/utils/dimensions.ts

```diff
@@ -109,217 +64,101 @@ function getAreaDimensions(insight: IInsightDefinition): IDimension[] {
-    const stack = insightBucket(insight, BucketNames.STACK);
-
-    if (!stack || bucketIsEmpty(stack)) {
-        return [
-            {
-                itemIdentifiers: viewAttributes.map(attributeLocalId),
-            },
-            {
-                itemIdentifiers: [MeasureGroupIdentifier],
-            },
-        ];
-    }
+    const viewByAttributes = safeBucketAttributes(insight, BucketNames.VIEW);
+    const stackByAttributes = safeBucketAttributes(insight, BucketNames.STACK);
 
-    return [
-        {
-            itemIdentifiers: viewAttributes.map(attributeLocalId),
-        },
-        {
-            itemIdentifiers: [...bucketAttributes(stack).map(attributeLocalId), MeasureGroupIdentifier],
-        },
-    ];
+    return stackByAttributes.length
+        ? newTwoDimensional(viewByAttributes, [...stackByAttributes, MeasureGroupIdentifier])
+        : newTwoDimensional(viewByAttributes, [MeasureGroupIdentifier]);
 }
 
 function getBubbleDimensions(insight: IInsightDefinition): IDimension[] {
-    const view = insightBucket(insight, BucketNames.VIEW);
-    const viewAttributes = view ? bucketAttributes(view) : [];
-    const stack = insightBucket(insight, BucketNames.STACK);
-
-    if (!stack || bucketIsEmpty(stack)) {
-        return [
-            {
-                itemIdentifiers: viewAttributes.map(attributeLocalId),
-            },
-            {
-                itemIdentifiers: [MeasureGroupIdentifier],
-            },
-        ];
-    }
+    const viewByAttributes = safeBucketAttributes(insight, BucketNames.VIEW);
+    const stackByAttributes = safeBucketAttributes(insight, BucketNames.STACK);
 
-    return [
-        {
-            itemIdentifiers: [
-                ...viewAttributes.map(attributeLocalId),
-                ...bucketAttributes(stack).map(attributeLocalId),
-            ],
-        },
-        {
-            itemIdentifiers: [MeasureGroupIdentifier],
-        },
-    ];
+    return newTwoDimensional([...viewByAttributes, ...stackByAttributes], [MeasureGroupIdentifier]);
 }
 
 /**
- * generateDimensions
- * is a function that generates dimensions based on buckets and visualization objects.
- * WARNING: It duplicates logic from pluggable visualizations.
- *          Remove once react components support pluggable visualizations.
+ * Generates dimensions based on buckets and visualization objects.
+ *
  * @param insight - insight being visualized
  * @param type - visualization type string
  * @internal
  */
 export function generateDimensions(insight: IInsightDefinition, type: VisType): IDimension[] {
     switch (type) {
-        case VisualizationTypes.TABLE: {
+        case VisualizationTypes.TABLE:
             return getPivotTableDimensions(insight);
-        }
+
         case VisualizationTypes.PIE:
         case VisualizationTypes.DONUT:
-        case VisualizationTypes.FUNNEL: {
+        case VisualizationTypes.FUNNEL:
             return getPieOrDonutDimensions(insight);
```

Comment:
could you rename getPieOrDonutDimensions now its not clear that dimensions are valid also for FUNNEL

Reply:
updated :)